scripts/inject-relup.escript

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env escript
+
+%% This script injects implicit relup instructions for emqx applications.
+
+-mode(compile).
+
+-define(ERROR(FORMAT, ARGS), io:format(standard_error, "[inject-relup] " ++ FORMAT ++ "~n", ARGS)).
+-define(INFO(FORMAT, ARGS), io:format(user, "[inject-relup] " ++ FORMAT ++ "~n", ARGS)).
+
+usage() ->
+    "Usage: " ++ escript:script_name() ++ " <path-to-release-dir> <release-vsn>".
+
+main([RelRootDir, CurrRelVsn]) ->
+    case filelib:is_dir(filename:join([RelRootDir, "releases"])) andalso
+         filelib:is_dir(filename:join([RelRootDir, "lib"])) of
+        true ->
+            EmqxAppVsns = get_emqx_app_vsns(RelRootDir),
+            ok = inject_relup_file(RelRootDir, CurrRelVsn, EmqxAppVsns);
+        false ->
+            ?ERROR("not a valid root dir of release: ~p, for example: _build/emqx/rel/emqx",
+                [RelRootDir]),
+            erlang:halt(1)
+    end;
+main(_Args) ->
+    ?ERROR("~s", [usage()]),
+    erlang:halt(1).
+
+inject_relup_file(RelRootDir, CurrRelVsn, EmqxAppVsns) ->
+    RelupFile = filename:join([RelRootDir, "releases", CurrRelVsn, "relup"]),
+    inject_file(RelupFile, EmqxAppVsns).
+
+inject_file(File, EmqxAppVsns) ->
+    case file:script(File) of
+        {ok, {CurrRelVsn, UpVsnRUs, DnVsnRUs}} ->
+            ?INFO("injecting instructions to: ~p", [File]),
+            UpdatedContent = {CurrRelVsn,
+                inject_relup_instrs(up, EmqxAppVsns, CurrRelVsn, UpVsnRUs),
+                inject_relup_instrs(down, EmqxAppVsns, CurrRelVsn, DnVsnRUs)},
+            file:write_file(File, term_to_text(UpdatedContent));
+        {ok, _BadFormat} ->
+            ?ERROR("bad formatted relup file: ~p", [File]),
+            error({bad_relup_format, File});
+        {error, enoent} ->
+            ?INFO("relup file not found: ~p", [File]),
+            ok;
+        {error, Reason} ->
+            ?ERROR("read relup file ~p failed: ~p", [File, Reason]),
+            error({read_relup_error, Reason})
+    end.
+
+inject_relup_instrs(Type, EmqxAppVsns, CurrRelVsn, RUs) ->
+    lists:map(fun
+        ({Vsn, "(relup-injected) " ++ _ = Desc, Instrs}) -> %% already injected
+            {Vsn, Desc, Instrs};
+        ({Vsn, Desc, Instrs}) ->
+            {Vsn, "(relup-injected) " ++ Desc,
+                append_emqx_relup_instrs(Type, EmqxAppVsns, CurrRelVsn, Vsn, Instrs)}
+    end, RUs).
+
+%% The `{apply, emqx_relup, post_release_upgrade, []}` will be appended to the end of
+%% the instruction lists.
+append_emqx_relup_instrs(up, EmqxAppVsns, CurrRelVsn, FromRelVsn, Instrs) ->
+    {EmqxVsn, true} = maps:get(CurrRelVsn, EmqxAppVsns),
+    Extra = #{}, %% we may need some extended args
+    %% we have to put 'load_object_code' before 'point_of_no_return'
+    %% so here we simply put it to the beginning
+    Instrs0 = [ {load_object_code, {emqx, EmqxVsn, [emqx_relup]}}
```

Comment:
question: this instruction will be duplicated if the module is added to `emqx.appup.src`, is it ok to be loaded twice ?

if it's ok, please add `emqx_app` module here as well.
if not, we'll need to add a ci step to ensure both `emqx_relup` and `emqx_app` are always added to to `emqx.appup.src`.

Reply:
For duplicated instructions, it is OK if every `load` instruction has a predefined `load_object_code`. But if they are not appear in pairs, e.g. there're 2 `load_object_code` but 3 `load` after that, the release upgrade will crash. 

I'll put emqx_app here and also do more checks about the instructions.